### PR TITLE
Foundation distillation: add response diagnostics and audited DINOv3 experiments

### PR DESCRIPTION
**Pull request by gao-666**
## Summary

This PR adds reusable, fail-closed Foundation response diagnostics and their
synthetic contracts, together with a compact and auditable DINOv3-S →
YOLO-Master-N distillation study.

The research result is intentionally bounded: the final matched-compute
three-seed comparison found **No detectable change** in clean detection
performance. The normalized response objective is therefore **not proposed
as a default training strategy**.

## What changed

- Added deterministic Foundation response-field utilities with:
  - frozen perturbation selection and noise seeding;
  - resume-stable logical batch indexing;
  - BatchNorm buffer preservation;
  - fail-closed FP32 cosine checks;
  - teacher-detached response loss.
- Added synthetic response contracts and regression tests.
- Added a compact research map, frozen final protocol, formal recommendation,
  and B24 recovery disclosure.
- Added four byte-identical final result tables plus an offline provenance
  index.

No trainer, queue, probe, raw batch audit, model weights, or dataset assets are
included in this PR.

## Primary result

Final matched-compute comparison:

- mean paired delta: **-0.000475**
- paired 95% t CI: **[-0.008120910, +0.007170910]**
- decision: **No detectable change**

This is not an equivalence claim; the interval still includes meaningful
positive and negative effects.

The secondary response-gap endpoint showed aggregate Mechanism Support, but
that result does not override the primary detection endpoint and is not a
causal mediation claim.

## Recovery disclosure

One B24 formal attempt was manually interrupted and archived. The retained
six-arm analysis uses a fresh epoch-0 restart under the same frozen
configuration.

`automatic_retries = 0` refers to queue policy only; it does not mean no manual
recovery occurred.

## Evidence

The original scientific freeze remains unchanged and is not published because
historical records contain host-local privacy-sensitive paths.

A sanitized derivative evidence package is publicly available at the immutable
commit:

`fedc360e73cdbfeaed930b17365458f632b1abc8`

The public package does not inherit the original research Git history.
Protected scientific result/protocol files were copied byte-for-byte; other
records were sanitized according to the documented privacy rules.

See:

- `experiments/rhino_d2/README.md`
- `experiments/rhino_d2/RESEARCH_MAP.md`
- `experiments/rhino_d2/EVIDENCE_INDEX.json`
- `experiments/rhino_d2/DINOV3_P2_FORMAL_RECOMMENDATION.md`
- `experiments/rhino_d2/P2_07_RECOVERY_DISCLOSURE.md`

## Validation

- 43 Foundation response tests passed.
- 6 offline evidence-index tests passed.
- 49 scoped tests passed in total.
- 10 Foundation response doctests passed.
- Scoped changed-code Ruff/format checks passed.
- The four committed result tables remain byte-identical to the frozen source
  and the public evidence package.

Repository-wide lint/format is not clean and is **not** claimed as passing.
Those unrelated findings were intentionally not expanded into this PR.

## Scope and recommendation

Foundation response diagnostics are opt-in and introduce no default inference
behavior change.

The reusable response/audit infrastructure is proposed for review. The
experiment-only normalized response training objective is **not** included as
a default training strategy.

The current D2 research route is terminal; this PR does not authorize or
propose additional alpha/loss/seed searching.

### CODE REVIEW

File: experiments/rhino_d2/P2_07_RECOVERY_DISCLOSURE.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+Public access: [public derivative of this document](https://github.com/gao-666/YOLO-Master/blob/fedc360e73cdbfeaed930b17365458f632b1abc8/experiments/rhino_d2/P2_07_RECOVERY_DISCLOSURE.md). Original document identity remains local at packaging SHA `781a6f148a46d59a89ea40806de8e10eb7b98941`. The public package is a sanitized derivative, not the original Git history; its PROVENANCE.json distinguishes byte-preserved files from path-sanitized records. Local-only references below are not public links. Import edits: this notice and pinned link destinations only; archived figures are linked rather than embedded. Scientific text and numbers are unchanged. The protocol's original implementation is distinct from post-freeze public-API hardening in integration Commit 1 `3a62b6f480e2855c3034f9788f69b804b81a5b56`.
+
+---
+
+# P2-07 / B24 人工中断恢复披露
+
+本页是事后整理，不冒充事前注册或当时的授权记录；不改写最终科学锚点 e3e11d0。
+
+原始隔离目录：`E:/2026YOLO/P2-07-interrupted-B24-20260909-224418`。
```
[the remark this answers is not in the document]
Comment (gao-666) [reply experiments/rhino_d2/P2_07_RECOVERY_DISCLOSURE.md]: Fixed in the latest review-follow-up commits (`c1328ef`, `a10ad29`). This is integration/public-API hardening only; frozen scientific results were not regenerated or modified.

File: ultralytics/nn/foundation/response.py [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+            f"{batch_index_within_epoch} >= {num_batches_per_epoch}."
+        )
+    return epoch_index * num_batches_per_epoch + batch_index_within_epoch
+
+
+def _validated_normalized_path(path: str | Path) -> str:
+    """Validate an already-normalized portable image path used by the digest contract."""
+    value = str(path)
+    if not value or "\0" in value or "\\" in value:
+        raise ValueError(f"normalized_image_path must be a non-empty POSIX-style path, got {value!r}.")
```
[the remark this answers is not in the document]
Comment (gao-666) [reply ultralytics/nn/foundation/response.py]: Fixed in the latest review-follow-up commits (`c1328ef`, `a10ad29`). Non-canonical and absolute paths are now rejected fail-closed, with regression coverage. Frozen scientific results were not regenerated or modified.

File: ultralytics/nn/foundation/response.py [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+    return output
+
+
+def tensor_sha256(tensor: torch.Tensor) -> str:
+    """Hash one finite contiguous CPU tensor including shape and dtype metadata."""
+    if not isinstance(tensor, torch.Tensor) or not torch.isfinite(tensor).all():
+        raise ValueError("tensor_sha256 requires a finite torch.Tensor.")
+    value = tensor.detach().contiguous().cpu()
+    header = f"{tuple(value.shape)}|{value.dtype}|".encode()
+    return hashlib.sha256(header + value.numpy().tobytes()).hexdigest()
```
[the remark this answers is not in the document]
Comment (gao-666) [reply ultralytics/nn/foundation/response.py]: Fixed in the latest review-follow-up commits (`c1328ef`, `a10ad29`). BF16 tensors are now hashed through a dtype-independent byte view, with regression coverage for BF16 determinism and the existing FP32 digest contract. Frozen scientific results were not regenerated or modified.

File: ultralytics/nn/foundation/response.py [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+    "GLOBAL_BATCH_INDEX_VERSION",
+    "RESPONSE_FIELD_CONDITIONS",
+    "RESPONSE_FIELD_PAYLOAD_VERSION",
+    "BatchNormBufferSnapshot",
+    "ResponseFieldCondition",
+    "apply_response_field_condition_batch",
+    "build_response_field_paired_view",
+    "logical_global_batch_index",
+    "preserve_batchnorm_buffers",
+    "response_field_condition",
```
[the remark this answers is not in the document]
Comment (gao-666) [reply ultralytics/nn/foundation/response.py:634]: Fixed in 2541fd0. The exported response/BN/digest helpers now carry runnable doctest examples. `python -m pytest --doctest-modules ultralytics/nn/foundation/response.py --override-ini addopts= -q` passes 10/10 doctests. This is integration/public-API hardening only; frozen scientific results were not regenerated or modified.

**Comment by isLinXu** [comment]
LGTM
